### DISCUSSION TITLE
Can FDTD.AddEdges2Grid be used with CSXCAD.CSPrimitives.CSPrimCylindricalShell and CSXCAD.CSPrimitives.CSPrimWire?

### DISCUSSION THREAD

**Discussion by oberstet** [thliebig/openEMS-Project#139] [Q&A Usage] [answered]
Hi all,

I'm a newbie, and currently I am struggling with adding a mesh for my geometry.

I am using `CSXCAD.CSPrimitives.CSPrimCylindricalShell` and `CSXCAD.CSPrimitives.CSPrimWire`, but I can't get `FDTD.AddEdges2Grid` working.

I _does_ work with `CSXCAD.CSPrimitives.CSPrimBox` though, which can be seen in my current mesh:

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/9f05a0cf-4d72-4729-8fc6-bd7e9eebd13a)

----

How do I auto-create a mesh for non-rectangular objects?

I've made sure the mesh resolution is fine enough, and tried various things like:

```python
    # FDTD.AddEdges2Grid(dirs='xyz', properties=helix1, metal_edge_res=mesh_res / 2)
    # FDTD.AddEdges2Grid(primitives=helix1, dirs='xyz', properties=helix1, metal_edge_res=mesh_res / 2)
    # FDTD.AddEdges2Grid(primitives=helix1, dirs='xyz', metal_edge_res=mesh_res / 2)
    # FDTD.AddEdges2Grid(primitives=helix1, dirs='all', metal_edge_res=mesh_res / 2)
    # FDTD.AddEdges2Grid(primitives=helix1, dirs='all')
    # FDTD.AddEdges2Grid(primitives=[helix1], dirs='all')
    # FDTD.AddEdges2Grid(primitives=[helix1], dirs='all', metal_edge_res=0.01)
    FDTD.AddEdges2Grid(properties=[helix1], dirs='all', metal_edge_res=0.01)
```

None of these work, there are no mesh lines added =(

----

Of course I could brute force it, and manually create sufficient number of lines at a resolution fine enough to "dissect" each individual `CSPrimCylindricalShell` ... not sure if this is what I'm supposed to do though .. maybe I'm just missing something?

**Comment by LubomirJagos** [comment, marked as answer]
This is how AddEdges2Grid looks inside
```python
def AddEdges2Grid(self, dirs, primitives=None, properties=None, **kw):
	""" AddEdges2Grid(primitives, dirs, **kw)

	Add the edges of the given primitives to the FDTD grid.

	:param dirs: primitives -- one or more primitives
	:param dirs: str -- 'x','y','z' or 'xy', 'yz' or 'xyz' or 'all'
	"""
	csx = self.GetCSX()
	if csx is None:
		raise Exception('AddEdges2Grid: Unable to access CSX!')
	prim_list = []
	if primitives is not None and  type(primitives) is not list:
		prim_list.append(primitives)
	elif primitives is not None:
		prim_list += primitives

	if properties is not None and  type(properties) is not list:
		prim_list += properties.GetAllPrimitives()
	elif properties is not None:
		for prop in properties:
			prim_list += prop.GetAllPrimitives()

	grid = csx.GetGrid()
	for prim in prim_list:
		hint = automesh.mesh_hint_from_primitive(prim, dirs, **kw)
		if hint is None:
			continue
		for n in range(3):
			if hint[n] is None:
				continue
			grid.AddLine(n, hint[n])
```

these are that hints what are called
```python
def mesh_hint_from_primitive(primitive, dirs, **kw):
    if primitive.GetType() is CSPrimitives.POINT:
        return mesh_hint_from_point(primitive, dirs, **kw)
    if primitive.GetType() is CSPrimitives.BOX:
        return mesh_hint_from_box(primitive, dirs, **kw)
    else:
        return None

def mesh_hint_from_point(point, dirs, **kw):
    """ mesh_hint_from_point(point, dirs)

    Get a grid hint for the coordinates of the point.

    :param dirs: str -- 'x','y','z' or 'xy', 'yz' or 'xyz' or 'all'
    :param mesh: combine mesh hint to existing mesh
    :returns: (3,) list of mesh hints
    """
    hint = [None, None, None]
    coord = point.GetCoord()
    for ny in GetMultiDirs(dirs):
        hint[ny] = [coord[ny],]
    if 'mesh' in kw:
        return mesh_combine(hint, kw['mesh'])
    return hint

def mesh_hint_from_box(box, dirs, **kw):
    """ mesh_hint_from_box(box, dirs, metal_edge_res=None, **kw)

    Get a grid hint for the edges of the given box with an an optional 2D metal
    edge resolution.

    :param dirs: str -- 'x','y','z' or 'xy', 'yz' or 'xyz' or 'all'
    :param metal_edge_res: float -- 2D flat edge resolution
    :param up_dir: bool -- Enable upper edge
    :param down_dir: bool -- Enable lower edge
    :param mesh: combine mesh hint to existing mesh
    :returns: (3,) list of mesh hints
    """
    metal_edge_res = kw.get('metal_edge_res', None)
    up_dir   = kw.get('up_dir'  , True)
    down_dir = kw.get('down_dir', True)

    if metal_edge_res is None:
        mer = 0
    else:
        mer = np.array([-1.0, 2.0])/3 * metal_edge_res
    if box.HasTransform():
        sys.stderr.write('FDTD::automesh: Warning, cannot add edges to grid with transformations enabled\n')
        return
    hint = [None, None, None]
    start = np.fmin(box.GetStart(), box.GetStop())
    stop  = np.fmax(box.GetStart(), box.GetStop())
    for ny in GetMultiDirs(dirs):
        hint[ny] = []
        if metal_edge_res is not None and stop[ny]-start[ny]>metal_edge_res:
            if down_dir:
                hint[ny].append(start[ny]-mer[0])
                hint[ny].append(start[ny]-mer[1])
            if up_dir:
                hint[ny].append(stop[ny]+mer[0])
                hint[ny].append(stop[ny]+mer[1])
        elif stop[ny]-start[ny]:
            if down_dir:
                hint[ny].append(start[ny])
            if up_dir:
                hint[ny].append(stop[ny])
        else:
            hint[ny].append(start[ny])
    if 'mesh' in kw:
        return mesh_combine(hint, kw['mesh'])
    return hint
```

so it just used boundaries of model, it doesn't have any information about your shape inside box, I WOULD STRONGLY NOT RECOMMEND USING THIS
instead as you state use fine enough mesh to intersect your model with gridlines and then run simulation like
  openEMS <xml file> --no-simulation --debug-PEC
  
open that generated file in paraview and inspect if metal cells are generated right you will see there mesh, this part is kind a manual work, yes you can try to automatize it somehow but first got your simulation working, also put there Efield monitors and inspect in paraview if field is propagate through your metal structure to be sure you have no defects in your structure

at this point you got from programming to engineering simulation :D :D :D like coding in python is fun, but real working with simulation is much more fun :)

**Reply by oberstet** [reply]
aaahh ..

```python
    else:
        return None
```

that explains it;)

np, I go for the manual route.

**Comment by LubomirJagos** [comment]
just for preview, you should see in paraview PEC mesh like this, this is for discone antenna in cylindrical coords, in cartesian there will be orthogonal wires used instead
![279485123-cdf99bfa-794d-40d2-a73d-efefc20f1c15](https://github.com/thliebig/openEMS-Project/assets/18229825/9d0eebc9-0ca8-40bb-ac8f-6827e899735f)

**Comment by oberstet** [comment]
>  but real working with simulation is much more fun :)

yeah!! I think I'm almost there.

once I got the mesh, I "only" am missing

* some way (port? lumpedport? plus some excitation. plus whatever)
* this fancy NF2FF transforming thing .. not sure why, but I guess I need it
* actually plot or what "how good is the antenna"

finally

* correct geometry (my problem with binops etc)
* materials/models for heatsink, battery and human hand

so there is still work to do ... should be doable in a couple of days

but in any case: thank you so much for your guidance!!! I mean, I can work hard to learn walking on new terrain myself, but if I walk in the wrong direction or that road is ultimately a dead end ... that would kinda suck .. I am lazy and lack patience;)

**Reply by LubomirJagos** [reply]
for your case:
    - lumped port in Z direction, gaussian excitation at f0=working freq, fc=simulation bandwith/2
    - NF2FF field monitor cube around antenna, don't make it too tight, it center should be also like in phase center of antenna but whatever, don't care too much about this
    - S11 parameter, at work frequency it should be <-10dB    
    - not sure if you did your geometry right since you are using copper traces and on picture it seems like there is huge space between trace and substrate, this will make your simulation unvalid, not real
    - materials are easy, but human hand should be defined as debeye material take example from MRI coil but I think there is special file for material permitivity and permeability so would no be useful, I will not even model hand, like there is not much influence as for EM field it is transparent (depends on case but my guess would be it is transparent)
    
like if you post here your KiCAD file I can show you magic how to do this in half hour instead of couple days :)

**Reply by LubomirJagos** [reply]
that NF2FF is near field -> far field, it calculates directional antenna gain that's that fancy graph how antennas radiate energy into space, since you are modelling helical antenna (sort of, your ground is strange) my expectation would be it will be torus with maximum alongside your spiral, minimum in antenna axis

**Comment by oberstet** [comment]
> like if you post here your KiCAD file I can show you magic how to do this in half hour instead of couple days :)

omg, that'd fantastic for sure! if only "seeing" any picture/plot with some simulation result for the antenna (kinda frustrating to do so much work for me without yet seeing any results)

here you go:

https://github.com/oberstet/subghz-dual-antenna/blob/main/subghz-dual-antenna.kicad_pcb

this repo is just a quick, semi-chaotic snapshot of the portions rgd the antenna from my main repo for the device project

the antenna is intended for:

*446MHz, 223MHz and 892MHz, as well as some MHz around those frequencies, and mostly broadband in between the whole range 200-950MHz, and primarily horizontally working, in all directions*

**Reply by oberstet** [reply]
plus: the current geometries in KiCad are based on my semi-informed guess work totally! not optimized, or even verified in any way, neither by simulation nor real-world test. it _should_ be resonant "somewhere subghz" I would assume .. but even that isn't verified

**Reply by oberstet** [reply]
btw, above repo also has my initial python experiments using openEMS https://github.com/oberstet/subghz-dual-antenna/blob/main/ems-model/subghz_dual_planar_helical_antenna.py ... this is experimental code, slight chaos, not cleaned up or properly using classes or what ...  just a quick hack

**Reply by oberstet** [reply]
finally, it also has FreeCAD exported from KiCad (my manually drawn hackish antenna) https://github.com/oberstet/subghz-dual-antenna/blob/main/subghz-dual-antenna.FCStd

**Comment by oberstet** [comment]
@LubomirJagos Hi there! did you achieve any results?

also, rgd meshing in general .. not sure .. this is what I have now .. does it make sense / is that what I'm supposed to do?

![grafik](https://github.com/thliebig/openEMS-Project/assets/233340/ea2dfdbf-cf5e-4132-9358-e8165a871789)

maybe I need a couple of more mesh lines in z-axis .. not sure ... also, rgd "model complexity / run-time": since I have zero clue, my current numbers of mesh lines as can be seen in the screenshot (though I'm not sure how to read it): is that "low" or "high"

again, sorry for all the dumb questions;) any help greatly appreciated!

**Reply by oberstet** [reply]
IOW: does my simulation run from a total of this many cells?

```
>>> 18*148*6
15984
```

and: is that "high" or "low"? I can then tweak accordingly ..
